src/System.Management.Automation/engine/regex.cs

```diff
@@ -204,9 +204,10 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
                 char ch = pattern[i];
 
                 //
-                // if it is a wildcard char, escape it
+                // if it is a wildcard char or escape char, escape it
                 //
-                if (IsWildcardChar(ch) && !charsNotToEscape.Contains(ch))
+                if ((IsWildcardChar(ch) || ch == escapeChar) &&
+                    !charsNotToEscape.Contains(ch))
```

Comment:
Why we always should escape the escape char? What scenario don't work without fix?

Reply:
1. When we accidentally created a file named ``` `[test`]2 ``` due to the incorrect handling of wildcard character, we cannot remove it using the -Path parameter with the tab completed name.
2. When I escaped string A and get B, and try to unescape B, I would expected that A will be returned, but it is not when A contains escape character.
3. The wildcard pattern parser always treat escape character as escape character.

Reply:
My understanding is that we should test if a file/directory/path exists before trying escape. If exist then return the path. Then continue with escaping wildcard chars. Then continue with expansion wildcards if -Path.
Also what if there is some special symbols in the path? Seems we should process them in cycle.

Reply:
Okay, I know what is wrong here. This part is to fix the `WildcardPattern::Escape` function itself, which is not necessary related to the file system.
When `WildcardPattern::Get` parse a pattern string, it always treats the escape character as escape character of an escape character. That means, ``` `` ``` will be parsed to `` ` ``.
And so is the `WildcardPattern::Unescape` method, `[WildcardPattern]::Unescape('``')` will return `` ` `` instead of ``` `` ```.
But, ``[WildcardPattern]::Escape('`')`` does not escape the escape character, which conflicts with the others.

Reply:
@kwkam I see your point.
We have the public API
```c#
public static string Escape(string pattern)
```
So the change is a breaking change and it is not desirable to change this behavior until  it is  established that this is a significant bug.

Reply:
@iSazonov Yes, we should be careful with changes in public API, but I doubt if this is not a significant bug. For example, there is a list of text:
```
`[HIGH] patch 1`
`[MED] update 1`
`[LOW] patch 1`
`[MED] patch 2`
`[HIGH] hotfix 3`
`[LOW] patch 2`
`[LOW] hotfix 3`
```
Now, I want to search for `` `[LOW] patch 1` `` and `` `[LOW] patch 2` ``, and I am using WildcardPattern:
```PowerShell
$prefix = '`[LOW] patch '
$pattern = '[1-2]'
$suffix = '`'
$sarg = [WildcardPattern]::Escape($prefix) + $pattern + [WildcardPattern]::Escape($suffix)
$opts = [Management.Automation.WildcardOptions]::Compiled
$matcher = [WildcardPattern]::Get($sarg, $opts)
$ListOfText | ? {$matcher.IsMatch($_)}
```
No result will be return because of the inconsistency between the Escape method and the WildcardPattern parser. (And this is also the cause of the issue in auto-completed filename which contains escape character)

Reply:
Thanks for good example!

We need PowerShell Committee conclusion.
/cc @SteveL-MSFT